File: python/sdk/jit_proxy/jitter/jitter_shotgun.py
import asyncio

from typing import Any, Coroutine

from solders.pubkey import Pubkey # type: ignore

from driftpy.drift_client import DriftClient
from driftpy.auction_subscriber.auction_subscriber import AuctionSubscriber
from driftpy.types import is_variant, UserAccount, Order, PostOnlyParams
from driftpy.accounts.get_accounts import get_user_stats_account

from jit_proxy.jitter.base_jitter import BaseJitter
from jit_proxy.jit_proxy_client import JitIxParams, JitProxyClient


class JitterShotgun(BaseJitter):
    def __init__(
        self,
        drift_client: DriftClient,
        auction_subscriber: AuctionSubscriber,
        jit_proxy_client: JitProxyClient,
        verbose: bool,
    ):
        super().__init__(drift_client, auction_subscriber, jit_proxy_client, verbose)

    async def subscribe(self):
        await super().subscribe()

    async def create_try_fill(
        self,
        taker: UserAccount,
        taker_key: Pubkey,
        taker_stats_key: Pubkey,
        order: Order,
        order_sig: str,
    ) -> Coroutine[Any, Any, None]:
        self.logger.info("JitterShotgun: Creating Try Fill")

        async def try_fill():
            taker_stats = await get_user_stats_account(
                self.drift_client.program, taker.authority
            )

            referrer_info = self.get_referrer_info(taker_stats)

            for i in range(order.auction_duration):
                params = (
                    self.perp_params.get(order.market_index)
                    if is_variant(order.market_type, "Perp")
                    else self.spot_params.get(order.market_index)
                )

                if params is None:
                    self.ongoing_auctions.pop(order_sig)
                    return

                self.logger.info(f"Trying to fill {order_sig} -> Attempt: {i + 1}")

                try:
                    if params.max_position == 0 and params.min_position == 0:
                        break
            
                    sig = await self.jit_proxy_client.jit(
                        JitIxParams(
                            taker_key,
                            taker_stats_key,
                            taker,
                            order.order_id,
                            params.max_position,
                            params.min_position,
                            params.bid,
                            params.ask,
                            params.price_type,
                            referrer_info,
                            params.sub_account_id,
                            PostOnlyParams.MustPostOnly(),
                        )
                    )

                    self.logger.info(f"Filled Order: {order_sig}")
                    self.logger.info(f"Signature: {sig}")
                    await asyncio.sleep(10)  # sleep for 10 seconds
                    del self.ongoing_auctions[order_sig]
                    return
                except Exception as e:
                    self.logger.error(f"Failed to fill Order: {order_sig}")
                    if "0x1770" in str(e) or "0x1771" in str(e):
                        self.logger.error(f"Order: {order_sig} does not cross params yet, retrying")
                    elif "0x1779" in str(e):
                        self.logger.error(f"Order: {order_sig} could not fill, retrying")
                    elif "0x1793" in str(e):
                        self.logger.error("Oracle invalid, retrying")
                    elif "0x1772" in str(e):
                        self.logger.error(f"Order: {order_sig} already filled")
                        # we don't want to retry if the order is filled
                        break
                    else:
                        self.logger.error(f"Error: {e}")
                        await asyncio.sleep(10)  # sleep for 10 seconds
                        del self.ongoing_auctions[order_sig]
                        return
            if order_sig in self.ongoing_auctions:
                del self.ongoing_auctions[order_sig]

        return await try_fill()

File: ts/sdk/src/jitter/jitterShotgun.ts
import { JitProxyClient } from '../jitProxyClient';
import { PublicKey } from '@solana/web3.js';
import {
	AuctionSubscriber,
	DriftClient,
	Order,
	PostOnlyParams,
	SwiftOrderSubscriber,
	UserAccount,
	UserStatsMap,
} from '@drift-labs/sdk';
import { BaseJitter } from './baseJitter';
import { SignedSwiftOrderParams } from '@drift-labs/sdk/lib/node/swift/types';

export class JitterShotgun extends BaseJitter {
	constructor({
		auctionSubscriber,
		jitProxyClient,
		driftClient,
		userStatsMap,
		swiftOrderSubscriber,
	}: {
		driftClient: DriftClient;
		auctionSubscriber: AuctionSubscriber;
		jitProxyClient: JitProxyClient;
		userStatsMap?: UserStatsMap;
		swiftOrderSubscriber?: SwiftOrderSubscriber;
	}) {
		super({
			auctionSubscriber,
			jitProxyClient,
			driftClient,
			userStatsMap,
			swiftOrderSubscriber,
		});
	}

	createTryFill(
		taker: UserAccount,
		takerKey: PublicKey,
		takerStatsKey: PublicKey,
		order: Order,
		orderSignature: string
	): () => Promise<void> {
		return async () => {
			let i = 0;

			const takerStats = await this.userStatsMap.mustGet(
				taker.authority.toString()
			);
			const referrerInfo = takerStats.getReferrerInfo();

			// assumes each preflight simulation takes ~1 slot
			while (i < order.auctionDuration) {
				const params = this.perpParams.get(order.marketIndex);
				if (!params) {
					this.deleteOnGoingAuction(orderSignature);
					return;
				}

				const txParams = {
					computeUnits: this.computeUnits,
					computeUnitsPrice: this.computeUnitsPrice,
				};

				console.log(`Trying to fill ${orderSignature}`);
				try {
					const { txSig } = await this.jitProxyClient.jit(
						{
							takerKey,
							takerStatsKey,
							taker,
							takerOrderId: order.orderId,
							maxPosition: params.maxPosition,
							minPosition: params.minPosition,
							bid: params.bid,
							ask: params.ask,
							postOnly: params.postOnlyParams ?? PostOnlyParams.MUST_POST_ONLY,
							priceType: params.priceType,
							referrerInfo,
							subAccountId: params.subAccountId,
						},
						txParams
					);

					console.log(
						`Successfully sent tx for ${orderSignature} txSig ${txSig}`
					);
					await sleep(10000);
					this.deleteOnGoingAuction(orderSignature);
					return;
				} catch (e) {
					console.error(`Failed to fill ${orderSignature}`);
					if (e.message.includes('0x1770') || e.message.includes('0x1771')) {
						console.log('Order does not cross params yet, retrying');
					} else if (e.message.includes('0x1779')) {
						console.log('Order could not fill');
					} else if (e.message.includes('0x1793')) {
						console.log('Oracle invalid, retrying');
					} else {
						await sleep(10000);
						this.deleteOnGoingAuction(orderSignature);
						return;
					}
				}
				i++;
			}

			this.deleteOnGoingAuction(orderSignature);
		};
	}

	createTrySwiftFill(
		authorityToUse: PublicKey,
		signedSwiftOrderParams: SignedSwiftOrderParams,
		uuid: Uint8Array,
		taker: UserAccount,
		takerKey: PublicKey,
		takerStatsKey: PublicKey,
		order: Order,
		orderSignature: string,
		marketIndex: number
	): () => Promise<void> {
		return async () => {
			let i = 0;

			const takerStats = await this.userStatsMap.mustGet(
				taker.authority.toString()
			);
			const referrerInfo = takerStats.getReferrerInfo();

			// assumes each preflight simulation takes ~1 slot
			while (i < order.auctionDuration) {
				const params = this.perpParams.get(order.marketIndex);
				if (!params) {
					this.deleteOnGoingAuction(orderSignature);
					return;
				}

				const txParams = {
					computeUnits: this.computeUnits,
					computeUnitsPrice: this.computeUnitsPrice,
				};

				console.log(`Trying to fill ${orderSignature}`);
				try {
					const { txSig } = await this.jitProxyClient.jitSwift(
						{
							takerKey,
							takerStatsKey,
							taker,
							takerOrderId: order.orderId,
							maxPosition: params.maxPosition,
							minPosition: params.minPosition,
							bid: params.bid,
							ask: params.ask,
							postOnly: params.postOnlyParams ?? PostOnlyParams.MUST_POST_ONLY,
							priceType: params.priceType,
							referrerInfo,
							subAccountId: params.subAccountId,
							authorityToUse,
							signedSwiftOrderParams,
							uuid,
							marketIndex,
						},
						txParams
					);

					console.log(
						`Successfully sent tx for ${orderSignature} txSig ${txSig}`
					);
					await sleep(10000);
					this.deleteOnGoingAuction(orderSignature);
					return;
				} catch (e) {
					console.error(`Failed to fill ${orderSignature}`);
					if (e.message.includes('0x1770') || e.message.includes('0x1771')) {
						console.log('Order does not cross params yet, retrying');
					} else if (e.message.includes('0x1779')) {
						console.log('Order could not fill');
					} else if (e.message.includes('0x1793')) {
						console.log('Oracle invalid, retrying');
					} else {
						await sleep(10000);
						this.deleteOnGoingAuction(orderSignature);
						return;
					}
				}
				i++;
			}

			this.deleteOnGoingAuction(orderSignature);
		};
	}
}

function sleep(ms: number): Promise<void> {
	return new Promise((resolve) => setTimeout(resolve, ms));
}
